### DISCUSSION TITLE
tracel-xtask in Cargo.toml automatically changes Cargo.lock

### DISCUSSION THREAD

**Discussion by tiruka** [tracel-ai/burn#2259] [General]
Thank you for developing burn.

It is not a bug, but I believe it should be modified. 

I know you are recently done with refactoring and separating xtask package. Now the tracel-xtask is born and it is imported as follows in the root Cargo.toml.

https://github.com/tracel-ai/burn/blob/a9f941d403686abdb5ff2f3bea80219ecd6228a9/Cargo.toml#L164

It updates Cargo.lock in the local users automatically because cargo is looking for latest packages based on Semantic Versioning, so patch version will be updated. 
When making a PR to main, it is possible to make conflicts in Cargo.lock. I also know tracel-xtask will be frequently updated.

My suggestion is pin the version to the patch (X.Y.Z) instead of `"~1.0`. 

I would like to hear developers opinions.

Thank you.

**Comment by syl20bnr** [comment]
This is a convenience trade-off, as you mentioned `tracel-xtask` can change a lot and ideally we would like to avoid micro-managing its version  in all our repos which can be tedious. So this is why the version patch is not pinned. It gives us the flexibility to choose whether we automatically dispatch changes to all our repos or we make a breaking change that requires explicit update of the `Cargo.toml` file.

We made the decision to include `Cargo.lock` in the Burn repository to increase reproducibility and avoid a whole class of dependencies issues but it comes with trade-offs as well. It is fine to have conflicts on this file in a PR though, in this case a rebase of the PR branch fixes it.

**Comment by tiruka** [comment]
Thank you for the explanation. I understand the trade-offs involved in not pinning the patch version of tracel-xtask and how this provides flexibility in managing updates across repositories. Your idea also makes sense.

Regarding the conflicts in Cargo.lock, I understand you allow them.

Thanks.
